### DISCUSSION TITLE
preloadとeager_loadの違いと使い分け

### DISCUSSION THREAD

**Discussion by yuyakawai0411** [kakeru-one/web-engineering-roundtable#50] [General]
### 背景
O/Rマッパーを使うことで起きるトラブルの代表的な例がN+1問題である。そのN+1問題を解決するために、preloadとeager_loadを使うが、それぞれの違いは何か？、どのように使い分けるべきか？
https://www.slideshare.net/slideshow/how-topreventorm-troubles/42902000

### preloadとeager_loadの違い
| 種類 | 特徴
| --- | --- |
| preload | 指定したアソシエーションを2回のクエリに分けて取得し、キャッシュする |
| eager_load | 指定したアソシエーションをleft outer joinで取得し、キャッシュする |

**Comment by kakeru-one** [comment]
データを絞り込まないときの、eager_loadとpreloadの使い分けは？
https://tech.itandi.co.jp/entry/2023/03/14/180000

**Reply by kakeru-one** [reply]
## どっちの方が速い？
- joinする計算コストが高いのは、テーブルのレコード量が多いとき
- preloadが遅いのは、Network I/Oに時間がかかる場合。

結論、データ量によって変わるので、使い分けはケースバイケースなのかなと思った。

**Reply by yuyakawai0411** [reply]
> 親と子テーブルが１：多や多：多の関係の時はpreloadで、 親と子テーブルが多：１の場合や関連先のテーブルの要素で絞り込みを行う場合は..

https://tech.itandi.co.jp/entry/2023/03/14/180000

取得した右側のテーブルのデータ量が多ければ、joinするのに時間がかかるため、preloadの方がいい。取得した右側のテーブルのデータ量が少なければ、SQLを発行する回数が少ないearger_loadの方がいいと思う。
